Update zc143ac72mipi co5300 driver and display driver example

KATE-WANG-NXP opened the pull request:
1. Add user configuration of buffer address alignment and update area pitch's alignment for display driver sample.
2. Add board specific configuration for mimxrt700_evk lvgl examples. The mimxrt700_evk uses DC8000 to drive all panels which requires a 64-byte align of buffer address and stride.
3. Update CO5300 panel driver
3.1 Fix wrong backlight pin in driver overlay
3.2 Remove the power-on pin configuration in code and binding, and add mipi display panel regulator in panel overlay instead. Set 'regulator-boot-on' to true means the power-on pin will be enabled during system boot.
3.3 Remove 'last_known_framebuffer' from panel data structure. It is not used anywhere
3.4 Fix bug in 'co5300_set_pixel_format' function.
3.5 Fix the issue that the panel does not support start coordinates and the width/height of the updated area being odd value.
       Solution: In panel driver, maintain a full screen-sized buffer, its address and pitch alignment is configurable in device tree and shall be compliant with the display controller's requirements. It can be placed in RAM or if the RAM space is not enough it can also be placed in other memory resion. When there is a frame update request, the updated area will be first filled to the buffer, if the area's size or coordinate is odd, adjust the value so the real updated area covers the requested updated area, then use this buffer to send pixel to panel, this can ensure the updated area's size and coordinate are always even.

EmilioCBen commented on drivers/display/display_co5300.c:
```diff
@@ -104,27 +100,34 @@ static int co5300_write(const struct device *dev,
+	 * area is row 6~472 and line 0~466. So adjust coordinates accordingly.
+	 */
 	/* First two bytes are starting X coordinate */
-	start_xpos = x;
```
Should we update the width and height of the panel in the shield overlay instead of adding this runtime calculations?

uLipe commented on drivers/display/display_co5300.c:
```diff
@@ -173,6 +175,12 @@ static int co5300_write(const struct device *dev,
+		if (framebuffer_size == 0U) {
+			break;
+		}
+
```
Is it guaranteed that a potential underflow will not occur here?

KATE-WANG-NXP replied:
Hi I updated the driver to use an internal maintained buffer now so I think there will not be such issue.

uLipe reviewed (commented):
I just have a question.

JarmouniA commented on samples/drivers/display/Kconfig:
```diff
@@ -0,0 +1,16 @@
+
+source "Kconfig.zephyr"
+
+config BUFFER_ADDR_ALIGN
```
```suggestion
config SAMPLE_BUFFER_ADDR_ALIGN
```

JarmouniA commented on samples/drivers/display/boards/mimxrt700_evk_mimxrt798s_cm33_cpu0.conf:
```diff
@@ -0,0 +1,8 @@
+#
+
+CONFIG_BUFFER_ADDR_ALIGN=64
+CONFIG_PITCH_ALIGN=64
```
Are these params dependent on the board or the display panel?

JarmouniA commented on samples/modules/lvgl/demos/boards/mimxrt700_evk_mimxrt798s_cm33_cpu0.conf:
```diff
@@ -0,0 +1,10 @@
+CONFIG_LV_Z_MEM_POOL_SIZE=81920
+# RT700 uses DC8000 to drive panel which requires a 64 byte align of buffer address and stride.
+CONFIG_LV_DRAW_BUF_STRIDE_ALIGN=64
+CONFIG_LV_DRAW_BUF_ALIGN=64
```
the LV mempool size is app-specific, so should stay here.

KATE-WANG-NXP committed:
samples: drivers: display: Add user configuration

Add user configuration of buffer address alignment and update area pitch
alignment.

Signed-off-by: Kate Wang <yumeng.wang@nxp.com>

KATE-WANG-NXP committed:
samples: lvgl: Add board specific configuration for mimxrt700_evk

The mimxrt700_evk uses DC8000 to drive panel which requires a 64
byte align of buffer address and stride.

Signed-off-by: Kate Wang <yumeng.wang@nxp.com>

KATE-WANG-NXP commented on samples/drivers/display/src/main.c:
```diff
@@ -241,8 +241,8 @@ int main(void)
-		rect_h = 1;
+		rect_w = 4;
+		rect_h = 2;
 	}
```
reverted

faxe1008 commented on drivers/display/Kconfig.co5300:
```diff
@@ -1,10 +1,41 @@
+
+config CO5300_EXTERNAL_FB_MEM
+	bool "Use external defined address for frame buffer"
+	default n
```
not needed for boolean

KATE-WANG-NXP commented on drivers/display/display_co5300.c:
```diff
@@ -104,27 +102,83 @@ static int co5300_write(const struct device *dev,
+		local_desc.height = desc->height;
+	}
+	local_desc.width = ROUND_UP(local_desc.width, 2U);
+	local_desc.height = ROUND_UP(local_desc.height, 2U);
```
Yes the update area start coordinate and the size shall all be even value, so if x = 1, width = 4, then x->0, width->5, and the width shall add another 1

zejiang0jason commented on drivers/display/display_co5300.c:
```diff
@@ -104,27 +102,83 @@ static int co5300_write(const struct device *dev,
+	local_desc.width = ROUND_UP(local_desc.width, 2U);
+	local_desc.height = ROUND_UP(local_desc.height, 2U);
+	local_desc.width = desc->width + 1;
+	local_desc.height = desc->height + 1;
```
why assign the `local_desc.width` and `local_desc.height` again?

zejiang0jason commented on drivers/display/display_co5300.c:
```diff
@@ -104,27 +102,83 @@ static int co5300_write(const struct device *dev,
+	local_desc.width = desc->width + 1;
+	local_desc.height = desc->height + 1;
+	local_desc.pitch = data->frame_pitch;
+	local_desc.frame_incomplete = desc->frame_incomplete;
```
can we only send out whole dirty region, when the `frame_incomplete` is false?

JarmouniA commented on drivers/display/display_rm67162.c:
unrelated to this PR?

KATE-WANG-NXP replied:
Move to another pr https://github.com/zephyrproject-rtos/zephyr/pull/99880

JarmouniA commented on drivers/mipi_dbi/mipi_dbi_nxp_dcnano_lcdif.c:
unrelated to this PR?

JarmouniA commented on boards/shields/zc143ac72mipi/boards/mimxrt700_evk_mimxrt798s_cm33_cpu0.overlay:
Please move board DT changes to a separate commit.

KATE-WANG-NXP committed:
boards: shield: Update shield support for zc143ac72mipi

The zc143ac72mipi panel can only accept update area
that has even size and coordinates, so the panel
driver was updated to maintain a buffer to collect
all dirty areas. This buffer shall have address and
pitch alignments which compliant to the board's display
driver's requirement, and can be placed outside of ram.
Update the board specific overlay to add such configurations.

Signed-off-by: Kate Wang <yumeng.wang@nxp.com>

JarmouniA commented on drivers/display/display_co5300.c:
```diff
@@ -230,37 +300,48 @@ static int co5300_set_pixel_format(const struct device *dev,
-		data->bytes_per_pixel = 2;
+		mipi_pixel_format = MIPI_DSI_PIXFMT_RGB565;
+		bytes_per_pixel = 2U;
+		cmd_params[0] = MIPI_DCS_ADDRESS_MODE_BGR;
```
When running the Display sample, is the output conform to what is described in the readme (color-wise)?
https://docs.zephyrproject.org/latest/samples/drivers/display/README.html

JarmouniA commented on dts/bindings/display/chipone,co5300.yaml:
```diff
@@ -29,8 +29,22 @@ properties:
+    type: phandle
+    description: |
+      External SRAM/SDRAM in which the panel's internal frame buffer will
+      be stored. If not defined, internal RAM will be used.
```
```suggestion
      Secondary RAM in which the panel controller's frame buffer will
      be stored. If not defined, the 'zephyr,ram' chosen memory will be used.
```

JarmouniA commented on drivers/display/display_co5300.c:
```diff
@@ -446,6 +527,21 @@ static DEVICE_API(display, co5300_api) = {
+			co5300_frame_buffer_##node_id[DT_INST_PROP(node_id, height) * 3U *	\
+			ROUND_UP(DT_INST_PROP(node_id, width),					\
+			DT_INST_PROP(node_id, pitch_align))]
+#define CO5300_FRAMEBUFFER(node_id) co5300_frame_buffer_##node_id
```
```suggestion

#define CO5300_FRAMEBUFFER(node_id) co5300_frame_buffer_##node_id
```

KATE-WANG-NXP commented on drivers/display/display_co5300.c:
```diff
@@ -446,6 +527,21 @@ static DEVICE_API(display, co5300_api) = {
 	.set_orientation = co5300_set_orientation,
 };
 
+#if DT_INST_NODE_HAS_PROP(0, ext_ram)
```
Yes, I tried to use instance but failed to find a reference, do you have an example on how to pass the instance in a #if?

KATE-WANG-NXP commented on drivers/display/display_co5300.c:
```diff
@@ -446,6 +527,21 @@ static DEVICE_API(display, co5300_api) = {
+
+#define CO5300_FRAMEBUFFER_DECL(node_id)							\
+	/* frame buffer aligned to cache line width for optimal cache flushing */		\
+	FRAME_BUFFER_SECTION static uint8_t __aligned(DT_INST_PROP(node_id, addr_align))	\
```
Sorry typo

JarmouniA commented on dts/bindings/display/chipone,co5300.yaml:
```diff
@@ -29,8 +29,22 @@ properties:
+    default: 1
+    description: |
+      The CO5300 maintains an internal frame buffer. Its address must be
+      compliant with the display controller's requirement.
```
```suggestion
      The CO5300 driver maintains an internal frame buffer. This property
      serves for defining an alignment for its address that's compliant
      with the SoC's display controller's requirements.
```

KATE-WANG-NXP committed:
drivers: display: display_co5300: Update panel driver

1. Fix wrong backlight pin in driver overlay
2. Remove the power-on pin configuration in code and binding, and
add mipi display panel regulator in panel overlay instead. Set
regulator-boot-on' to true means the power-on pin will be enabled
uring system boot.
3. Remove 'last_known_framebuffer' from panel data structure. It is
not used anywhere
4. Fix bug in 'co5300_set_pixel_format' function.
5. Fix the issue that the panel does not support start coordinates
   and the width/height of the updated area being odd value.
   Solution: In panel driver, maintain a full screen-sized buffer,
   its address and pitch alignment is configurable in device tree
   and shall be compliant with the display controller's requirements.
   It can be placed in RAM or if the RAM space is not enough it can
   also be placed in other memory resion. When there is a frame
   update request, the updated area will be first filled to the
   buffer, if the area's size or coordinate is odd, adjust the value
   so the real updated area covers the requested updated area, then
   use this buffer to send pixel to panel, this can ensure the
   updated area's size and coordinate are always even.

Signed-off-by: Kate Wang <yumeng.wang@nxp.com>
